crates/ruff/src/registry.rs

```diff
@@ -23,6 +24,12 @@ impl Rule {
             .find(|rule| rule.noqa_code().suffix() == code)
             .ok_or(FromCodeError::Unknown)
     }
+
+    pub fn from_name(name: &str) -> Result<Self, FromCodeError> {
```

Comment:
I think there are some usages in `crates/ruff/src/checkers/noqa.rs` that need to be adjusted too (look for `if let Ok(rule) = Rule::from_code(code) {` and `let code = get_redirect_target(code).unwrap_or(code)`). In short, we need to audit all the usages of `noqa_code()`, in that file too.

Reply:
It may end up being useful to have separate wrapper types to make this all clearer (I haven't tried this myself but it's worth exploring), like:

```
pub struct RuleCodeId(String);
pub struct RuleNameId(String);
```

Which would ensure that we only compare `RuleCodeId` to `NoqaCode`, etc.

(In `crates/ruff/src/checkers/noqa.rs`, we often pass around the raw codes as `&str` because we don't yet _know_ if they're valid rule codes, and so we can't pass them around as `NoqaCode`.)

Reply:
I refactored the logic in `noqa.rs` a bit so we can pass around `Rule`s, which removes the ambiguity between a code and a name. Please let me know how it looks.